File: rawasi/src/services/chatService.js
// src/services/chatService.js
import { supabase } from "../lib/supabase";

//------------------------------- Start of Proxy addition -------------------------------
import RealDatabaseHandler from "../services/database/RealDatabaseHandler";
import ProxyDatabaseHandler from "../services/database/ProxyDatabaseHandler";
//-------------------------------------------- End of Proxy addition -------------------------------

const meId = async () => {
  const { data: { user } } = await supabase.auth.getUser();
  if (!user) throw new Error("Not authenticated");
  return user.id;
};

export const chatService = {

  // Create or get conversation
  getOrCreateConversation: async (otherId, projectId) => {
    const me = await meId();
    const [a, b] = me < otherId ? [me, otherId] : [otherId, me];

    //------------------------------- Start of Proxy addition -------------------------------
    const db = new ProxyDatabaseHandler(new RealDatabaseHandler(), true);
    //-------------------------------------------- End of Proxy addition -------------------------------

    const { data, error } = await db.upsertConversation({
      project_id: projectId,
      a,
      b
    });

    if (error) throw error;
    return data;
  },

  // List all conversations for current user
  listConversations: async () => {
    const me = await meId();

    //------------------------------- Start of Proxy addition -------------------------------
    const db = new ProxyDatabaseHandler(new RealDatabaseHandler(), true);
    //-------------------------------------------- End of Proxy addition -------------------------------

    const { data: convs, error } = await db.getConversations(me);
    if (error) throw error;

    const rows = await Promise.all(
      convs.map(async (c) => {
        const otherId = c.a === me ? c.b : c.a;

        const [{ data: other }, { data: last }] = await Promise.all([
          supabase.from("profiles").select("id,name,role").eq("id", otherId).single(),
          supabase
            .from("messages")
            .select("body,created_at,sender,is_read")
            .eq("conversation_id", c.id)
            .order("created_at", { ascending: false })
            .limit(1)
        ]);

        return { ...c, other, last: last?.[0] || null };
      })
    );

    return rows;
  },

  // Fetch all messages inside a conversation
  fetchMessages: async (conversationId) => {

    //------------------------------- Start of Proxy addition -------------------------------
    const db = new ProxyDatabaseHandler(new RealDatabaseHandler(), true);
    //-------------------------------------------- End of Proxy addition -------------------------------

    const { data, error } = await db.fetchMessages(conversationId);
    if (error) throw error;
    return data;
  },

  // Send a new message
  send: async (conversationId, text) => {
    const me = await meId();

    //------------------------------- Start of Proxy addition -------------------------------
    const db = new ProxyDatabaseHandler(new RealDatabaseHandler(), true);
    //-------------------------------------------- End of Proxy addition -------------------------------

    const { error } = await db.sendMessage(conversationId, me, text);
    if (error) throw error;
  },

  // Mark all messages in conversation as read
  markRead: async (conversationId) => {
    const me = await meId();

    //------------------------------- Start of Proxy addition -------------------------------
    const db = new ProxyDatabaseHandler(new RealDatabaseHandler(), true);
    //-------------------------------------------- End of Proxy addition -------------------------------

    await db.markMessageRead(conversationId, me);
  },

  // Supabase realtime subscription (Proxy not needed)
  subscribe: (conversationId, onNew) => {
    const channel = supabase
      .channel("chat-" + conversationId)
      .on("postgres_changes",
        { event: "INSERT", schema: "public", table: "messages", filter: `conversation_id=eq.${conversationId}` },
        (payload) => onNew(payload.new)
      )
      .subscribe();

    return () => supabase.removeChannel(channel);
  }
};

File: rawasi/src/test/chat_service.py
# chat_service.py
from supabase_auth.errors import AuthApiError 

class ChatService:
    def __init__(self, supabase_client):
        self.client = supabase_client

    async def _get_me_id(self):
        """دالة مساعدة لجلب هوية المستخدم الحالي"""
        try:
            user_response = await self.client.auth.get_user()
            user = user_response.user
            if not user:
                raise Exception("Not authenticated")
            return user.id
        except AuthApiError:  # <-- هذا السطر سيعمل الآن بشكل صحيح
            raise Exception("Not authenticated")

    async def send(self, conversation_id: str, text: str):
        """إرسال رسالة جديدة"""
        me_id = await self._get_me_id()
        try:
            await self.client.from_("messages").insert({
                "conversation_id": conversation_id,
                "sender": me_id,
                "body": text
            }).execute()
        except Exception as e:
            print(f"Error sending message: {e}")
            raise

    async def get_or_create_conversation(self, other_id: str, project_id: str):
        """جلب أو إنشاء محادثة"""
        me_id = await self._get_me_id()
        
        # فرز الهويات لضمان الترتيب الصحيح
        a, b = (me_id, other_id) if me_id < other_id else (other_id, me_id)
        
        try:
            # Upsert يضمن إنشاء المحادثة إذا لم تكن موجودة
            response = await self.client.from_("conversations").upsert(
                {"project_id": project_id, "a": a, "b": b},
                on_conflict="project_id,a,b"
            ).select("id").single().execute()
            
            return response.data # { 'id': ... }
        except Exception as e:
            print(f"Error getting/creating conversation: {e}")
            raise